### PR TITLE
Update CCI extensions to understand new C# lang features

### PR DESCRIPTION
**Pull request by weshaggard**
With this change our tools like genapi and apicompat will understand

in params
ref readonly returns
readonly structs
ref structs

Also added apicompat rules to ensure the new annotations are compatible

Addresses https://github.com/dotnet/corefx/issues/25039.

cc @jkotas @jaredpar @VSadov 

These updates were used to find the incompatible annotations and regenerate ref files in PR https://github.com/dotnet/corefx/pull/26175.

### CODE REVIEW

File: src/ApiCompat/Rules/Compat/TypeCannotChangeClassification.cs [hunk 1]
```diff
@@ -20,8 +21,19 @@ public override DifferenceType Diff(IDifferences differences, ITypeDefinition im
...
+                    "Type '{0}' is a '{1}' in the implementation but is a '{2}' in the contract.",
                     impl.FullName(), implObjType, contractObjType);
+
+                return DifferenceType.Changed;
+            }
+
+            if (impl.Attributes.HasIsReadOnlyAttribute() && !contract.Attributes.HasIsReadOnlyAttribute())
+            {
+                differences.AddIncompatibleDifference(this,
+                    "Type '{0}' is marked as readonly in the implementation so it must also be marked readonly in the contract.",
```
Comment (jkotas) [inline src/ApiCompat/Rules/Compat/TypeCannotChangeClassification.cs]: I think this should be other way around: 
- readonly in implementation only is ok
- readonly in contract only is not ok

File: src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Attributes.cs [hunk 2]
```diff
@@ -387,7 +387,24 @@ private static bool ExcludeSpecialAttribute(ICustomAttribute c)
...
                 case "System.Runtime.CompilerServices.ExtensionAttribute": return true;
                 case "System.Runtime.CompilerServices.DynamicAttribute": return true;
-            }
+                case "System.Runtime.CompilerServices.IsByRefLikeAttribute": return true;
+                case "System.Runtime.CompilerServices.IsReadOnlyAttribute": return true;
+                case "System.ObsoleteAttribute":
+                    {
+                        var arg = c.Arguments.OfType<IMetadataConstant>().FirstOrDefault();
+
+                        if (arg != null && arg.Value is string)
```
Comment (justinvp) [inline src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Attributes.cs]: Nit: Could be shortened to:

```c#
case "System.ObsoleteAttribute":
    return c.Arguments.OfType<IMetadataConstant>().FirstOrDefault()?.Value is string argValue &&
        argValue == "Types with embedded references are not supported in this version of your compiler.";
```

File: src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Attributes.cs [hunk 3]
```diff
@@ -387,7 +387,24 @@ private static bool ExcludeSpecialAttribute(ICustomAttribute c)
...
+                            string argValue = (string)arg.Value;
+
+                            if (argValue == "Types with embedded references are not supported in this version of your compiler.")
+                            {
+                                return true;
+                            }
+                        }
+                        break;
+                    }
+    }
```
Comment (justinvp) [inline src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Attributes.cs]: Nit: Indentation of this `}` looks off.

File: src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Attributes.cs (hunk 2 above)
Comment (weshaggard) [reply src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Attributes.cs]: I'm not going turn it all the way into a statement like you have but I will use a few of the shortened code fragments.

File: src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Attributes.cs (hunk 3 above)
Comment (weshaggard) [reply src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Attributes.cs]: fixed.

File: src/ApiCompat/Rules/Compat/TypeCannotChangeClassification.cs (hunk 1 above)
Comment (weshaggard) [reply src/ApiCompat/Rules/Compat/TypeCannotChangeClassification.cs]: Fixed.

File: src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs [hunk 4]
```diff
@@ -213,19 +225,31 @@ private void WriteMethodBody(IMethodDefinition method)
...
-            else if (!TypeHelper.TypesAreEquivalent(method.Type, method.ContainingTypeDefinition.PlatformType.SystemVoid))
-            {
-                WriteKeyword("throw null;");
+                Write("throw null; ");
             }
 
             WriteSymbol("}");
         }
 
+        private bool NeedsMethodBodyForCompilation(IMethodDefinition method)
```
Comment (jkotas) [inline src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs]: Nit: You can keep it simple and just emit `throw null` all the time. `throw null` vs. empty body has no impact on the binary size - the binary size will be same in both cases.

Comment (weshaggard) [reply src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs]: I guess that is true. I wasn't as concerned wit the binary size as much as I was the source code size. Ideally I would just be empty bodies all the time. Perhaps the new compiler /refonly flag might allow that I should check.

Comment (VSadov) [reply src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs]: " => throw null; "    is not a lot of code :-) 

Allowing empty bodies (or no bodies at all - like in interfaces) was considered, but I am not sure it was done (yet?). 
CC: @jcouv

**Comment by VSadov** [comment]
CC:@jbevain @tmat  @jaredpar  - do we need to do similar changes in the context of cecil and/or metadata reader?

**Comment by tmat** [comment]
I don't think we would need to update Cecil, since it operates on IL and shouldn't have any code that would need to understand C# language features. 
What will need to be updated is decompilers like ILSpy. FYI @sharwell

**Comment by weshaggard** [comment]
test Innerloop Windows_NT Debug

**Pull request merged**